<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />




<title>tsRNAsearch Report</title>

<script src="CytC_vs_TotalRNA.Results-summary_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="CytC_vs_TotalRNA.Results-summary_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="CytC_vs_TotalRNA.Results-summary_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="CytC_vs_TotalRNA.Results-summary_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="CytC_vs_TotalRNA.Results-summary_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="CytC_vs_TotalRNA.Results-summary_files/navigation-1.1/tabsets.js"></script>
<link href="CytC_vs_TotalRNA.Results-summary_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="CytC_vs_TotalRNA.Results-summary_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="fluid-row" id="header">



<h1 class="title toc-ignore">tsRNAsearch Report</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#overview-of-tsrnasearch-analysis-comparing-cytc-versus-totalrna"><span class="toc-section-number">1</span> Overview of tsRNAsearch analysis comparing CytC versus TotalRNA</a><ul>
<li><a href="#quantities-of-ncrna-features-identified"><span class="toc-section-number">1.1</span> Quantities of ncRNA features identified</a><ul>
<li><a href="#proportions-of-ncrna-groups-identified-normalised-by-reads-per-million"><span class="toc-section-number">1.1.1</span> Proportions of ncRNA groups identified (normalised by reads per million)</a></li>
<li><a href="#number-of-raw-reads-identified-for-each-ncrna-group"><span class="toc-section-number">1.1.2</span> Number of raw reads identified for each ncRNA group</a></li>
</ul></li>
<li><a href="#venn-diagram-comparing-the-three-methods-used-deseq2-distribution-algorithm-cleavage-algorithm"><span class="toc-section-number">1.2</span> Venn diagram comparing the three methods used (DESeq2, distribution algorithm, cleavage algorithm)</a><ul>
<li><a href="#details-of-the-intersections-can-be-found-here"><span class="toc-section-number">1.2.1</span> Details of the intersections can be found here:</a></li>
<li><a href="#files-used-to-generate-the-venn-diagram-can-be-found-here"><span class="toc-section-number">1.2.2</span> Files used to generate the Venn Diagram can be found here:</a></li>
</ul></li>
<li><a href="#summary-report-table"><span class="toc-section-number">1.3</span> Summary Report Table</a></li>
<li><a href="#trna-read-alignment-length"><span class="toc-section-number">1.4</span> tRNA Read Alignment Length</a></li>
</ul></li>
<li><a href="#differential-expression-results-deseq2"><span class="toc-section-number">2</span> Differential Expression Results (DESeq2)</a><ul>
<li><a href="#overview-of-de-analysis"><span class="toc-section-number">2.1</span> Overview of DE analysis</a><ul>
<li><a href="#principal-component-analysis-pca-of-all-samples"><span class="toc-section-number">2.1.1</span> Principal Component Analysis (PCA) of all samples</a></li>
<li><a href="#sample-distance-matrix"><span class="toc-section-number">2.1.2</span> Sample distance matrix</a></li>
<li><a href="#volcano-plot-showing-differentially-expressed-genes"><span class="toc-section-number">2.1.3</span> Volcano plot showing differentially expressed genes</a></li>
</ul></li>
<li><a href="#features-identified"><span class="toc-section-number">2.2</span> Features identified</a><ul>
<li><a href="#differentially-expressed-trnas"><span class="toc-section-number">2.2.1</span> Differentially expressed tRNAs</a></li>
<li><a href="#coverage-plots-for-all-differentially-expressed-trnas"><span class="toc-section-number">2.2.2</span> Coverage plots for all differentially expressed tRNAs</a></li>
<li><a href="#differentially-expressed-ncrnas-other-than-trnas"><span class="toc-section-number">2.2.3</span> Differentially expressed ncRNAs (other than tRNAs)</a></li>
<li><a href="#coverage-plots-for-all-differentially-expressed-ncrnas"><span class="toc-section-number">2.2.4</span> Coverage plots for all differentially expressed ncRNAs</a></li>
</ul></li>
</ul></li>
<li><a href="#distribution-algorithm-results"><span class="toc-section-number">3</span> Distribution Algorithm Results</a><ul>
<li><a href="#feature-identified"><span class="toc-section-number">3.1</span> Feature identified</a><ul>
<li><a href="#highest-scoring-trna-fragments-identified-using-the-distribution-algorithm"><span class="toc-section-number">3.1.1</span> Highest scoring tRNA fragments identified using the distribution algorithm</a></li>
<li><a href="#coverage-plots-for-all-trnas-with-irregular-distribution-identified-by-the-distribution-algorithm"><span class="toc-section-number">3.1.2</span> Coverage plots for all tRNAs with irregular distribution identified by the distribution algorithm</a></li>
<li><a href="#highest-scoring-ncrna-fragments-identified-using-the-distribution-algorithm"><span class="toc-section-number">3.1.3</span> Highest scoring ncRNA fragments identified using the distribution algorithm</a></li>
<li><a href="#coverage-plots-for-all-snomirnas-with-irregular-distribution-identified-by-the-distribution-algorithm"><span class="toc-section-number">3.1.4</span> Coverage plots for all sno/miRNAs with irregular distribution identified by the distribution algorithm</a></li>
</ul></li>
</ul></li>
<li><a href="#cleavage-algorithm-results"><span class="toc-section-number">4</span> Cleavage Algorithm Results</a><ul>
<li><a href="#features-identified-1"><span class="toc-section-number">4.1</span> Features identified</a><ul>
<li><a href="#highest-scoring-trna-fragments-identified-using-the-cleavage-algorithm"><span class="toc-section-number">4.1.1</span> Highest scoring tRNA fragments identified using the cleavage algorithm</a></li>
<li><a href="#coverage-plots-for-all-trnas-with-irregular-5-vs-3-coverage-identified-by-the-cleavage-algorithm"><span class="toc-section-number">4.1.2</span> Coverage plots for all tRNAs with irregular 5’ vs 3’ coverage identified by the cleavage algorithm</a></li>
<li><a href="#highest-scoring-ncrna-fragments-identified-using-the-cleavage-algorithm"><span class="toc-section-number">4.1.3</span> Highest scoring ncRNA fragments identified using the cleavage algorithm</a></li>
<li><a href="#coverage-plots-for-all-snomirnas-with-irregular-5-vs-3-coverage-identified-by-the-cleavage-algorithm"><span class="toc-section-number">4.1.4</span> Coverage plots for all sno/miRNAs with irregular 5’ vs 3’ coverage identified by the cleavage algorithm</a></li>
</ul></li>
</ul></li>
<li><a href="#more-results"><span class="toc-section-number">5</span> More results:</a><ul>
<li><a href="#trna-fragments-identified-where-read-origin-accurate-only-to-isoacceptor-level"><span class="toc-section-number">5.0.1</span> tRNA fragments identified where read origin accurate only to isoacceptor level</a></li>
<li><a href="#other"><span class="toc-section-number">5.1</span> Other</a><ul>
<li><a href="#all-trna-coverage-plots-can-be-found-here"><span class="toc-section-number">5.1.1</span> All tRNA coverage plots can be found here:</a></li>
<li><a href="#run-information"><span class="toc-section-number">5.1.2</span> Run information</a></li>
</ul></li>
</ul></li>
</ul>
</div>

<div id="overview-of-tsrnasearch-analysis-comparing-cytc-versus-totalrna" class="section level1">
<h1><span class="header-section-number">1</span> Overview of tsRNAsearch analysis comparing CytC versus TotalRNA</h1>
<div id="quantities-of-ncrna-features-identified" class="section level2">
<h2><span class="header-section-number">1.1</span> Quantities of ncRNA features identified</h2>
<div id="proportions-of-ncrna-groups-identified-normalised-by-reads-per-million" class="section level3">
<h3><span class="header-section-number">1.1.1</span> Proportions of ncRNA groups identified (normalised by reads per million)</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_BarPlot_RPM-normalised.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="number-of-raw-reads-identified-for-each-ncrna-group" class="section level3">
<h3><span class="header-section-number">1.1.2</span> Number of raw reads identified for each ncRNA group</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_BarPlot_Raw-readcounts.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
</div>
<div id="venn-diagram-comparing-the-three-methods-used-deseq2-distribution-algorithm-cleavage-algorithm" class="section level2">
<h2><span class="header-section-number">1.2</span> Venn diagram comparing the three methods used (DESeq2, distribution algorithm, cleavage algorithm)</h2>
<p><img src="Plots/CytC_vs_TotalRNA_VennDiagram.png" width="600pt" height="600pt" style="display: block; margin: auto;" /></p>
</div>
</div>
<div id="summary-report-table" class="section level2">
<h2><span class="header-section-number">1.3</span> Summary Report Table</h2>
<table>
<caption>Top 20 results shown (sorted by distribution score)</caption>
<thead>
<tr class="header">
<th align="left">feature</th>
<th align="right">Fishers.method.pvalue</th>
<th align="right">distribution.score</th>
<th align="right">DESeq2.Log2FC</th>
<th align="right">DESeq2.pvalue</th>
<th align="right">DESeq2.padj</th>
<th align="right">cleavage.score</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">ENSMUSG00000084667 (Gm23212)</td>
<td align="right">0.00e+00</td>
<td align="right">474021.81</td>
<td align="right">6.813353</td>
<td align="right">0.0000001</td>
<td align="right">0.0000049</td>
<td align="right">1986907.81</td>
</tr>
<tr class="even">
<td align="left">ArgTCT</td>
<td align="right">0.00e+00</td>
<td align="right">195969.64</td>
<td align="right">0.704852</td>
<td align="right">0.0953823</td>
<td align="right">0.5779143</td>
<td align="right">205076512.63</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000077797 (Snord19)</td>
<td align="right">0.00e+00</td>
<td align="right">179005.74</td>
<td align="right">6.390443</td>
<td align="right">0.0000043</td>
<td align="right">0.0000940</td>
<td align="right">294860.83</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000064500 (Gm25296)</td>
<td align="right">0.00e+00</td>
<td align="right">150449.23</td>
<td align="right">6.048704</td>
<td align="right">0.0000253</td>
<td align="right">0.0004885</td>
<td align="right">3240661.06</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000080463 (Gm22489)</td>
<td align="right">0.00e+00</td>
<td align="right">136496.68</td>
<td align="right">5.210625</td>
<td align="right">0.0029302</td>
<td align="right">0.0392198</td>
<td align="right">344760.56</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000065126 (Snord104)</td>
<td align="right">0.00e+00</td>
<td align="right">133758.27</td>
<td align="right">5.079992</td>
<td align="right">0.0046642</td>
<td align="right">0.0541042</td>
<td align="right">983135.03</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000064780 (Snord95)</td>
<td align="right">0.00e+00</td>
<td align="right">133421.41</td>
<td align="right">4.868463</td>
<td align="right">0.0102711</td>
<td align="right">0.1116979</td>
<td align="right">1232681.11</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000077737 (Snord72)</td>
<td align="right">1.06e-05</td>
<td align="right">107327.81</td>
<td align="right">5.161541</td>
<td align="right">0.0044514</td>
<td align="right">0.0541042</td>
<td align="right">386338.11</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000065378 (Gm22744)</td>
<td align="right">0.00e+00</td>
<td align="right">83197.36</td>
<td align="right">5.205716</td>
<td align="right">0.0028431</td>
<td align="right">0.0392198</td>
<td align="right">417353.77</td>
</tr>
<tr class="even">
<td align="left">CysGCA</td>
<td align="right">0.00e+00</td>
<td align="right">77218.72</td>
<td align="right">-3.738328</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">20402930.28</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000065752 (Gm23344)</td>
<td align="right">0.00e+00</td>
<td align="right">70619.57</td>
<td align="right">4.466000</td>
<td align="right">0.0372228</td>
<td align="right">0.2943984</td>
<td align="right">189141.53</td>
</tr>
<tr class="even">
<td align="left">LysTTT</td>
<td align="right">1.57e-05</td>
<td align="right">65204.81</td>
<td align="right">4.488051</td>
<td align="right">0.0000000</td>
<td align="right">0.0000000</td>
<td align="right">10518645.88</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000065628 (Snord33)</td>
<td align="right">0.00e+00</td>
<td align="right">56191.66</td>
<td align="right">4.464234</td>
<td align="right">0.0369042</td>
<td align="right">0.2943984</td>
<td align="right">15844023.78</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000064721 (Gm25855)</td>
<td align="right">0.00e+00</td>
<td align="right">47250.34</td>
<td align="right">5.479416</td>
<td align="right">0.0008959</td>
<td align="right">0.0141709</td>
<td align="right">113270.48</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000065272 (Snord57)</td>
<td align="right">0.00e+00</td>
<td align="right">42456.96</td>
<td align="right">3.997330</td>
<td align="right">0.0979653</td>
<td align="right">0.5779143</td>
<td align="right">118131.89</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000077440 (Gm23130)</td>
<td align="right">3.00e-07</td>
<td align="right">36137.35</td>
<td align="right">3.853663</td>
<td align="right">0.1185328</td>
<td align="right">0.6653129</td>
<td align="right">125072.73</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000077625 (Snord4a)</td>
<td align="right">0.00e+00</td>
<td align="right">29540.88</td>
<td align="right">3.349669</td>
<td align="right">0.1958687</td>
<td align="right">0.8927857</td>
<td align="right">61271.49</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000065904 (Gm26109)</td>
<td align="right">0.00e+00</td>
<td align="right">27702.22</td>
<td align="right">3.133930</td>
<td align="right">0.2361876</td>
<td align="right">0.9175460</td>
<td align="right">285799.44</td>
</tr>
<tr class="odd">
<td align="left">ENSMUSG00000077527 (Gm25506)</td>
<td align="right">5.55e-05</td>
<td align="right">27290.77</td>
<td align="right">4.762864</td>
<td align="right">0.1968988</td>
<td align="right">0.8927857</td>
<td align="right">85382.35</td>
</tr>
<tr class="even">
<td align="left">ENSMUSG00000076361 (Mir182)</td>
<td align="right">2.00e-07</td>
<td align="right">26363.90</td>
<td align="right">3.332113</td>
<td align="right">0.2001071</td>
<td align="right">0.8927857</td>
<td align="right">19647024.76</td>
</tr>
</tbody>
</table>
</div>
<div id="trna-read-alignment-length" class="section level2">
<h2><span class="header-section-number">1.4</span> tRNA Read Alignment Length</h2>
<embed src="blob/master/additional-files/ExampleOutput/CytC_tRNA-read-alignment-lengths.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
<embed src="blob/master/additional-files/ExampleOutput/TotalRNA_tRNA-read-alignment-lengths.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
</div>
<div id="differential-expression-results-deseq2" class="section level1">
<h1><span class="header-section-number">2</span> Differential Expression Results (DESeq2)</h1>
<div id="overview-of-de-analysis" class="section level2">
<h2><span class="header-section-number">2.1</span> Overview of DE analysis</h2>
<div id="principal-component-analysis-pca-of-all-samples" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Principal Component Analysis (PCA) of all samples</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_tpm-PCA.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="sample-distance-matrix" class="section level3">
<h3><span class="header-section-number">2.1.2</span> Sample distance matrix</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_Distance-Matrix.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="volcano-plot-showing-differentially-expressed-genes" class="section level3">
<h3><span class="header-section-number">2.1.3</span> Volcano plot showing differentially expressed genes</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_VolcanoPlot.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
</div>
<div id="features-identified" class="section level2">
<h2><span class="header-section-number">2.2</span> Features identified</h2>
<div id="differentially-expressed-trnas" class="section level3">
<h3><span class="header-section-number">2.2.1</span> Differentially expressed tRNAs</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_tsRNAs.high-DE-negLog10_padj.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="coverage-plots-for-all-differentially-expressed-trnas" class="section level3">
<h3><span class="header-section-number">2.2.2</span> Coverage plots for all differentially expressed tRNAs</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_Features_Differentially-expressed-tsRNAs.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="differentially-expressed-ncrnas-other-than-trnas" class="section level3">
<h3><span class="header-section-number">2.2.3</span> Differentially expressed ncRNAs (other than tRNAs)</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_ncRNAs.high-DE-negLog10_padj.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="coverage-plots-for-all-differentially-expressed-ncrnas" class="section level3">
<h3><span class="header-section-number">2.2.4</span> Coverage plots for all differentially expressed ncRNAs</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_Features_Differentially-expressed-ncRNAs.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
</div>
<div id="distribution-algorithm-results" class="section level1">
<h1><span class="header-section-number">3</span> Distribution Algorithm Results</h1>
<div id="feature-identified" class="section level2">
<h2><span class="header-section-number">3.1</span> Feature identified</h2>
<div id="highest-scoring-trna-fragments-identified-using-the-distribution-algorithm" class="section level3">
<h3><span class="header-section-number">3.1.1</span> Highest scoring tRNA fragments identified using the distribution algorithm</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_tsRNAs.high-distribution-score.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="coverage-plots-for-all-trnas-with-irregular-distribution-identified-by-the-distribution-algorithm" class="section level3">
<h3><span class="header-section-number">3.1.2</span> Coverage plots for all tRNAs with irregular distribution identified by the distribution algorithm</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_Features_High-distribution-tsRNAs.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="highest-scoring-ncrna-fragments-identified-using-the-distribution-algorithm" class="section level3">
<h3><span class="header-section-number">3.1.3</span> Highest scoring ncRNA fragments identified using the distribution algorithm</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_ncRNAs.high-distribution-score.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="coverage-plots-for-all-snomirnas-with-irregular-distribution-identified-by-the-distribution-algorithm" class="section level3">
<h3><span class="header-section-number">3.1.4</span> Coverage plots for all sno/miRNAs with irregular distribution identified by the distribution algorithm</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_Features_High-distribution-ncRNAs.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
</div>
</div>
<div id="cleavage-algorithm-results" class="section level1">
<h1><span class="header-section-number">4</span> Cleavage Algorithm Results</h1>
<div id="features-identified-1" class="section level2">
<h2><span class="header-section-number">4.1</span> Features identified</h2>
<div id="highest-scoring-trna-fragments-identified-using-the-cleavage-algorithm" class="section level3">
<h3><span class="header-section-number">4.1.1</span> Highest scoring tRNA fragments identified using the cleavage algorithm</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_tsRNAs.high-cleavage-score.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="coverage-plots-for-all-trnas-with-irregular-5-vs-3-coverage-identified-by-the-cleavage-algorithm" class="section level3">
<h3><span class="header-section-number">4.1.2</span> Coverage plots for all tRNAs with irregular 5’ vs 3’ coverage identified by the cleavage algorithm</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_Features_Potentially-cleaved-tsRNAs.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="highest-scoring-ncrna-fragments-identified-using-the-cleavage-algorithm" class="section level3">
<h3><span class="header-section-number">4.1.3</span> Highest scoring ncRNA fragments identified using the cleavage algorithm</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_ncRNAs.high-cleavage-score.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="coverage-plots-for-all-snomirnas-with-irregular-5-vs-3-coverage-identified-by-the-cleavage-algorithm" class="section level3">
<h3><span class="header-section-number">4.1.4</span> Coverage plots for all sno/miRNAs with irregular 5’ vs 3’ coverage identified by the cleavage algorithm</h3>
<embed src="blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_Features_Potentially-cleaved-ncRNAs.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
</div>
</div>
<div id="more-results" class="section level1">
<h1><span class="header-section-number">5</span> More results:</h1>
<div id="trna-fragments-identified-where-read-origin-accurate-only-to-isoacceptor-level" class="section level3">
<h3><span class="header-section-number">5.0.1</span> tRNA fragments identified where read origin accurate only to isoacceptor level</h3>
<embed src="CytC_vs_TotalRNA_Features_tsRNAs-with-multimapping-reads.pdf" width="600pt" height="600pt" style="display: block; margin: auto;" type="application/pdf" />
</div>
<div id="other" class="section level2">
<h2><span class="header-section-number">5.1</span> Other</h2>
<div id="all-trna-coverage-plots-can-be-found-here" class="section level3">
<h3><span class="header-section-number">5.1.1</span> All tRNA coverage plots can be found here:</h3>
<ul>
<li>blob/master/additional-files/ExampleOutput/CytC_vs_TotalRNA_Features_All-tsRNAs.pdf</li>
</ul>
</div>
<div id="run-information" class="section level3">
<h3><span class="header-section-number">5.1.2</span> Run information</h3>
<div id="date-and-time" class="section level4">
<h4><span class="header-section-number">5.1.2.1</span> Date and time</h4>
<p>Pipeline initiated at Tue Nov 5 16:28:58 GMT 2019</p>
<p>Pipeline finished at Tue Nov 5 16:34:54 GMT 2019</p>
</div>
<div id="parameters-used" class="section level4">
<h4><span class="header-section-number">5.1.2.2</span> Parameters used</h4>
<p><br />Parameters: <br /> Species (-s): mouse <br /> Input directory containing fastq/fastq.gz files (-d): ExampleData/ <br /> Input fastq/fastq.gz file (-f): NA <br /> Experiment layout file (-e): additional-files/Example_ExperimentLayout.csv <br /> Output directory that tsRNAsearch will create and populate with results (-o): ExampleOutput <br /> Number of threads to use for the analysis (-t): 20 <br /> Plot all features (-A): no <br /> Skip pre-processing (-S): no <br /> Remove intermediate files (-R): yes</p>
</div>
<div id="version" class="section level4">
<h4><span class="header-section-number">5.1.2.3</span> Version</h4>
<p>tsRNAsearch version 0.12</p>
</div>
</div>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
